### DISCUSSION TITLE
How to install both Intel & Apple Silicon versions of formulae?

### DISCUSSION THREAD

**Discussion by MarkCallow** [Homebrew/discussions#2141] [Everyday usage] [answered]
[I spent 45 minutes looking over the documentation for answers to this and another 15 looking over discussions here.]

I want to be able to build both Intel and Apple Silicon versions of my apps in my CI system which runs on Intel Macs. Therefore I need to install both versions of my app's dependencies. After I install Brew - actually the CI provider already has Brew installed - how can I install both versions of the dependencies. Some specific questions:

1. Presumably the default when running on an Intel Mac is to install the formulae under `/usr/local` with symlinks to `/usr/local/bin`, etc.
2. How can I request Apple Silicon versions of formulae? The only relevant option I've seen is `--bottle-arch` but that talks about "optimizing for".
3. How can I find out which formulae are Apple Silicon ready?
4. If `--bottle-arch arm64` would indeed do what I need, will Brew install the package under `/opt/homebrew` or, because it is the Intel version of Brew running, will it put it under /usr/local.
5. Assuming the Apple Silicon version of a formula does get installed under `/opt/homebrew`, where will I find its binaries, `/opt/homebrew/bin`, etc.?

P.S. I do not understand the statements that sudo is not necessary. It is necessary on macOS when writing to `/usr/local` so at the vary least `brew` needs sudo to create its cellar directory and again whenever it needs to make links to `/usr/local/bin`, etc.

**Comment by SMillerDev** [comment]
None of this is possible, ARM macs can emulate x86 instructions but not the other way around.

---
> P.S. I do not understand the statements that sudo is not necessary. It is necessary on macOS when writing to /usr/local so at the vary least brew needs sudo to create its cellar directory and again whenever it needs to make links to /usr/local/bin, etc.

It's only needed for the installer because that will adjust the permissions of the relevant folders.

**Reply by MarkCallow** [reply]
> None of this is possible, ARM macs can emulate x86 instructions but not the other way around.


There is no need for any emulation of ARM instructions on x86 Macs to do what I'm wanting. You cross compile ARM binaries by specifying `-arch arm64` to clang. Then all you need is ARM versions of the libraries to link to in order to complete the build, which can likewise be cross-compiled, if installing from source.

**Reply by SMillerDev** [reply]
While that might be true, homebrew doesn't allow you to install anything you'll never be able to run anyway (we'd be a pretty annoying package manager if we did). So you can't install arm software on x86.

**Comment by alebcay** [comment, marked as answer]
> Presumably the default when running on an Intel Mac is to install the formulae under `/usr/local` with symlinks to `/usr/local/bin`, etc.

Yes, formulae are installed in a prefix `/usr/local/Cellar/<package_name>/<version_name>` and then symlinks are made into `/usr/local`, e.g. `/usr/local/Cellar/cmake/3.21.0/bin/cmake -> /usr/local/bin/cmake`. They're also symlinked into an unversioned "opt-prefix", `/usr/local/Cellar/cmake/3.21.0/bin/cmake -> /usr/local/opt/cmake/bin/cmake`.

> How can I request Apple Silicon versions of formulae? The only relevant option I've seen is `--bottle-arch` but that talks about "optimizing for".

You can `brew fetch <package> --bottle-tag <arch>`, e.g. `brew fetch cmake --bottle-tag arm64_big_sur`. This will fetch the precompiled copy for you. Note that the binaries/libraries inside the package are not immediately ready for use (Homebrew calls `install_name_tool` to perform some relocations for the target system on installation). You can download the bottle and then `brew install <path_to_bottle>` and Homebrew will naively install the package. Note that Homebrew will not check the architecture of the bottle and will install it into your system's `HOMEBREW_PREFIX`. (And all of this, for the record, is unsupported usage if you run into problems.)

> How can I find out which formulae are Apple Silicon ready?

There's no easy way to check this right now. `brew unbottled` (read `brew unbottled --help` for details) is available but this exists more for maintainers to work on building things that aren't natively on Apple Silicon yet.

> If `--bottle-arch arm64` would indeed do what I need, will Brew install the package under /opt/homebrew or, because it is the Intel version of Brew running, will it put it under /usr/local.

I don't think it would do what you need, but if it did, it would put it under `/usr/local`. Homebrew will always install formula into the `HOMEBREW_PREFIX` regardless of the bottle arch.

> Assuming the Apple Silicon version of a formula does get installed under `/opt/homebrew`, where will I find its binaries,` /opt/homebrew/bin`, etc.?

On Apple Silicon (Homebrew installed on an Apple Silicon system, not Homebrew installing an Apple Silicon package), formulae are installed in a prefix `/opt/homebrew/Cellar/<package_name>/<version_name>` and then symlinks are made into `/opt/homebrew`, e.g. `/opt/homebrew/Cellar/cmake/3.21.0/bin/cmake -> /opt/homebrew/bin/cmake`. They're also symlinked into an unversioned "opt-prefix", `/opt/homebrew/Cellar/cmake/3.21.0/bin/cmake -> /opt/homebrew/opt/cmake/bin/cmake`.

---

Generally, the use case you're asking for is unsupported. Right now it's not possible to install two different arches of the same formula in Homebrew without renaming one or the other. We've also considered shipping fat binaries/libraries in the past (back in the days of `ppc64`/`i386`/`x64`) but from what I've heard it was a road of pain. See also https://github.com/Homebrew/discussions/discussions/412 for discussion on why Homebrew has chosen not to ship Universal/fat binaries/libs, and see also https://github.com/Homebrew/discussions/discussions/16.

The libraries/software that Homebrew ships was never intended to be used/distributed outside of the Homebrew ecosystem, and due to the necessary core tool and CI changes that would be needed to support other uses (e.g. for redistribution by developers), I think it would be unlikely to happen.

Maybe a better place to get what you need would be `vcpkg` or `conan` package managers, which are geared more towards C/C++ developers for obtaining libraries suitable for distributing, rather than Homebrew which is geared more towards providing packages for end users.
